src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentRuntimeNodeWriter.cs

```diff
@@ -404,22 +404,49 @@ public override void WriteComponent(CodeRenderingContext context, ComponentInter
             }
             else
             {
+                var parameters = GetTypeInferenceMethodParameters(node.TypeInferenceNode);
+
+                // If this component is going to cascade any of its generic types, we have to split its type inference
+                // into two parts. First we call an inference method that captures all the parameters in local variables,
+                // then we use those to call the real type inference method that emits the component. The reason for this
+                // is so the captured variables can be used by descendants without re-evaluating the expressions.
+                CodeWriterExtensions.CSharpCodeWritingScope? typeInferenceCaptureScope = null;
+                if (node.Component.SuppliesCascadingGenericParameters())
+                {
+                    typeInferenceCaptureScope = context.CodeWriter.BuildScope();
+                    context.CodeWriter.Write(node.TypeInferenceNode.FullTypeName);
+                    context.CodeWriter.Write(".");
+                    context.CodeWriter.Write(node.TypeInferenceNode.MethodName);
+                    context.CodeWriter.Write("_CaptureParameters(");
+                    var isFirst = true;
+                    foreach (var parameter in parameters.Where(p => p.UsedForTypeInference))
+                    {
+                        if (isFirst)
+                        {
+                            isFirst = false;
+                        }
+                        else
+                        {
+                            context.CodeWriter.Write(", ");
+                        }
+
+                        WriteTypeInferenceMethodParameterInnards(context, parameter);
+                        context.CodeWriter.Write(", out var ");
+
+                        var variableName = $"__typeInferenceArg_{_scopeStack.Depth}_{parameter.ParameterName}";
+                        context.CodeWriter.Write(variableName);
+
+                        TrackCapturedCascadingGenericParameterVariable(node, parameter, variableName);
```

Comment:
nit: Consider a more explicit method name: `SubstituteProvidedGenericParameterValueWithCapturingVariable` or something along those lines.

Reply:
OK. I've gone with `UseCapturedCascadingGenericParameterVariable` because it's not just doing a one-time substitution - it's also keeping track of the variable for future use.